# Chapter 5

## Problem 1

Consider the problem, where we want to get the value of the Golden ratio, $\phi$.

The minimal polynomial, whose root is Golden ratio is $f(x) = x^2 - x - 1$.

Here is a Julia implementation of Newton's method in one dimension applied to the polynomial $f(x) = x^2 - x - 1$ with derivative $f'(x) = 2x - 1$.

In [6]:
function newton(f::Function, fp::Function, x::Real, tol::Real)
    (xold, xnew) = (x, Inf)
    for iteration = 1:100
        xnew = xold - f(xold) / fp(xold)
        if abs(f(xnew)) < tol
            return (xnew, iteration)
        end
        xold = xnew
    end
    return (xnew, 100)
end

newton (generic function with 1 method)

In [7]:
f(x) = x^2 - x - 1.0

f (generic function with 1 method)

In [8]:
fp(x) = 2x - 1.0

fp (generic function with 1 method)

In [9]:
(phi, iterations) = newton(f, fp, 1.0, 1e-14)

(1.618033988749895, 6)

In [10]:
f(phi)

0.0

In [24]:
(conjugate_phi, iterations) = newton(f, fp, 0.0, 1e-14)

(-0.6180339887498948, 6)

In [25]:
f(conjugate_phi)

-1.1102230246251565e-16

The value of the golden ratio $\phi$ as found by the above Julia code is 1.618033988749895 and its conjugate $\phi'$ is -0.6180339887498948. We can check our result using the fzero function of the Julia package Roots.jl.

In [13]:
using Roots

In [18]:
phi_roots = fzero(f, 1)

1.618033988749895

In [23]:
conjugate_phi_roots = fzero(f, 0)

-0.6180339887498949

In [22]:
phi - phi_roots

0.0

In [27]:
conjugate_phi - conjugate_phi_roots

1.1102230246251565e-16

The roots ($\phi, \phi'$) obtained using the fzero function from Roots.jl package is same as the ones obtained using Newton's method implementation.

## Problem 9

For $y>0$, the positive root of the equation $f(x) = \frac{1}{x^2} - y = 0$ is $\frac{1}{\sqrt y}$.

$f(x) = \frac{1}{x^2} - y$

$f'(x) = -2x^{-3} = \frac{-2}{x^3}$

Newton's iterates for finding the roots of the above function, $f(x) = 0$ :

$x_{n+1} = x_{n} - \frac{f(x_{n})}{f'(x_{n})}$

$\implies x_{n+1} = x_{n} - \frac{\frac{1}{x_n^2} - y}{-2x_n^{-3}}$

$\implies x_{n+1} = x_{n} + \frac{x_n^{3}}{2} (\frac{1}{x_n^2} - y)$

$\implies x_{n+1} = x_{n} + \frac{1}{2} (x_n - x_n^{3}y)$

$\implies x_{n+1} = \frac{1}{2} (3x_n - x_n^{3}y)$

$\implies x_{n+1} = \frac{x_n}{2} (3 - x_n^{2}y)$

$QED.$

Alternatively, $x = \frac{1}{\sqrt y}$ solves the equation $g(x) = yx^2 - 1 = 0$.

$g(x) = yx^2 - 1$

$g'(x) = 2yx$

Newton's iterates for finding the roots of the above function, $g(x) = 0$ :

$x_{n+1} = x_{n} - \frac{g(x_{n})}{g'(x_{n})}$

$\implies x_{n+1} = x_{n} - \frac{yx_n^2 - 1}{2yx_n}$

$\implies x_{n+1} = x_{n} - \frac{x_{n}}{2} + \frac{1}{2yx_n}$

$\implies x_{n+1} = \frac{1}{2} (x_{n} + \frac{1}{yx_n})$

$QED.$

The first scheme involves no reciprocals, but the second scheme has better convergence guarantees, which we prove next.

Consider for the second scheme, $x_n > 0$.

Using AM-GM Inequality, which states that for any two non-negative numbers a and b, the statement holds that

$\frac {a+b}{2} \geq \sqrt {ab}$ with equality if and only if a = b.

So, let $a = x_{n}$ and $b = \frac{1}{yx_n}$ in second scheme.

As $x_n > 0$ and $y > 0$ $\implies a>0$ and $b>0$.

Applying AM-GM Inequality, we get

$x_{n+1} = \frac{1}{2} (x_{n} + \frac{1}{yx_n}) \geq \sqrt {x_{n} \frac{1}{yx_n}} $

$ \implies x_{n+1} \geq \frac{1}{\sqrt y} \forall x_n > 0 $

$QED.$

Consider $x_n$ such that $x_{n} \geq \frac{1}{\sqrt y} > 0$.

$ \implies x^2_{n} \geq \frac{1}{y}$

$ \implies x_{n} \geq \frac{1}{yx_n}$ , as $x_n > 0$

$ \implies \frac{x_n}{2} \geq \frac{1}{2yx_n}$ , dividing by 2 on both sides.

$ \implies x_n \geq \frac{x_n}{2} + \frac{1}{2yx_n}$ , adding $\frac{x_n}{2}$ on both sides.

$ \implies x_n \geq x_{n+1}$, as $ x_{n+1} = \frac{1}{2} (x_{n} + \frac{1}{yx_n}) = \frac{x_n}{2} + \frac{1}{2yx_n}$.

$QED.$

Thus, $\forall x_n > 0, x_n \geq x_{n+1}$ and $ x_{n+1} \geq \frac{1}{\sqrt y} $.

So the sequence is monotonically decreasing and is lower bounded by $\frac{1}{\sqrt y}$ for all $x_n > 0$. Hence, if we take our initial estimate, $x_0 > 0$ , the global convergence is assured for the second scheme.

$QED.$